qa/test_template.sh

```diff
@@ -45,13 +45,56 @@ epilog=${epilog-:}
 # get the number of elements in `prolog` array
 numer_of_prolog_elms=${#prolog[@]}
 
+# turn on sanitizers
+enable_sanitizer() {
+    if [ -n "$DALI_ENABLE_SANITIZERS" ]; then
+        # supress leaks that are false positive or not related to DALI
+        export LSAN_OPTIONS=suppressions=$topdir/qa/leak.sup
+        export ASAN_OPTIONS=symbolize=1:protect_shadow_gap=0:log_path=sanitizer.log:start_deactivated=true:allocator_may_return_null=1::detect_leaks=1
+        export ASAN_SYMBOLIZER_PATH=$(which llvm-symbolizer)
+        # avoid python false positives
+        export PYTHONMALLOC=malloc
+    fi
+}
+
+# turn off sanitizer to avoid breaking any non-related system build-ins
+disable_sanitizer() {
+    if [ -n "$DALI_ENABLE_SANITIZERS" ]; then
+        export ASAN_OPTIONS=start_deactivated=true:detect_leaks=0
+        unset ASAN_SYMBOLIZER_PATH
+        unset PYTHONMALLOC
+    fi
+}
+
 # Wrap the test_body in a subshell, where we can safely execute it with `set -e`
 # and turn it off in current shell to intercept the error code
+# when sanitizers are on, do set +e to run all the test no matter what the result is
+# and collect as much of sanitizers output as possible
 test_body_wrapper() {(
-    set -e
+    if [ -n "$DALI_ENABLE_SANITIZERS" ]; then
+        set +e
+    else
+        set -e
+    fi
+    enable_sanitizer
```

Comment:
Might be a personal preference, but I think it'd read better if the if's were around the calls to enable_sanitizer/disable_sanitizer/process_sanitizers_logs and not inside the functions